### PR TITLE
try to reduce complexity of circuits.py

### PR DESCRIPTION
**Pull request by timmysilv**
Codefactor always complains of complex code due to circuits.py - I've made 3 changes:
- inline the `_ixdict` builder, it's only used in that once place
- inline the `propagate_shapes` helper, again, only used in that one place
- move `lookup_path` to a function that takes a Circuit, instead of it being an instance method

### CODE REVIEW

File: mrmustard/lab_dev/circuits.py [hunk 1]
```diff
@@ -624,3 +496,120 @@ def component_to_str(comp: CircuitComponent) -> list[str]:
             ret += "\n\n"
 
         return ret
+
+
+def lookup_path(circuit: Circuit) -> None:
```
Comment (apchytr) [inline mrmustard/lab_dev/circuits.py:501]: I'm not sure how I feel about it being a standalone method but maybe this should be in a `circuits_utils` folder? We could even include `_indices_dict` there

Comment (apchytr) [reply mrmustard/lab_dev/circuits.py:501]: Maybe even `component_to_str` should be in circuit_components_utils

Comment (apchytr) [reply mrmustard/lab_dev/circuits.py:501]: The point is that we have a lot of fairly large class methods that maybe make sense to be standalone util methods significantly cleaning up our files

**Comment by timmysilv** [comment]
better ways to go about this

**Pull request closed without merging**